### PR TITLE
Update -compat to support -compat 8.10

### PR DESCRIPTION
**Pull request by JasonGross**
**Kind:** infrastructure

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).

This PR institutes the policy that three `-compat` versions are supported on release branches, and four `-compat` versions are supported on master.  The `dev/tools/update-compat.py` script is updated to account for this policy, and now supports the `--master` flag (to be run on master) and the `--release` flag (to be used right before branching).

This is on top of #8634, and should be reviewed and merged once that PR is merged.

Note that this PR does not remove any compat notations.

### CODE REVIEW

**Review by ejgallego — APPROVED** [review]
My parts (`Flags` and `Coqargs`) look fine.

File: test-suite/bugs/closed/4798.v
Comment (Zimmi48) [reply test-suite/bugs/closed/4798.v]: This is because the compatibility notations are linked to `-compat 8.6` and removing compatibility with 8.6 should force us to remove the compatibility notations, but @silene is not OK to update Flocq in a hurry regarding these compatibility notations, so it's better to postpone the removal to 8.10 that way.

Comment (Zimmi48) [reply test-suite/bugs/closed/4798.v]: And BTW, this is part of #8634 anyways.

**Review by Zimmi48 — DISMISSED** [review]
I don't like the idea that we will remove the compatibility with previous version close to the release date because the whole point to do this kind of regular clean-up is to lighten the development of the new version. So in principle, I think this removal should happen ASAP after the branching.

So I liked your initial proposal (which was using the information on the current version and didn't distinguish between master and release) better but I still had one issue with it which is that it was blocking to allow updating the version number. However, this also shouldn't be the case. The version number should be bumped as soon as the previous version is branched off master. In fact, I plan to make the release process even more automated by having @coqbot do the branching automatically when one merges the PR bumping the version number in master.

So what I can propose is that as soon as the branching is done, @coqbot opens a PR updating the compatibility infrastructure by running the script. If this passes CI, it can be merged immediately, and if it doesn't (which is expected), then developers must ensure that projects in CI and the test-suite are fixed ASAP to allow merging the PR. This should be sufficient to avoid forgetting about this, and we can continue to hard-code the version number in the test-suite file.

File: test-suite/bugs/closed/4798.v
Comment (JasonGross) [reply test-suite/bugs/closed/4798.v]: No, that's not the reason here.  The reason is that this bug is testing the behavior of compat notations, and so we need to keep updating the compat version as we drop support for old versions.

**Comment by JasonGross** [comment]
Rebased to account for merge of #8634, and removed outdated review requests (people who were code owners for files in #8634 but not in this PR alone).

> I don't like the idea that we will remove the compatibility with previous version close to the release date because the whole point to do this kind of regular clean-up is to lighten the development of the new version. So in principle, I think this removal should happen ASAP after the branching.

I've changed "Shortly before the previous version is branched off master" to "Anytime before the previous version is branched off master".  Does this address your concerns?

> So I liked your initial proposal (which was using the information on the current version and didn't distinguish between master and release) better but I still had one issue with it which is that it was blocking to allow updating the version number. However, this also shouldn't be the case. The version number should be bumped as soon as the previous version is branched off master. In fact, I plan to make the release process even more automated by having @coqbot do the branching automatically when one merges the PR bumping the version number in master.

In this PR, I have separated out the compat updates into two steps: (1) permit `-compat NEW.VERSION`, and (2) remove `-compat VERY.OLD.VERSION`.  Step (1) is fully automated, and corresponds to passing the `--master` flag once the version number is bumped.  Since step (1) should never result in failures, I propose that we take it one step further, and have a script that bumps the version number and the magic numbers, etc, and runs `./dev/tools/update-compat.py --master --git-add`, and that this is the script used to construct the commit bumping the version number.  Then, @coqbot can branch on the commit right before the version number bump, and also prepare a PR on master that is the result of running `./dev/tools/update-compat.py --release --git-add`, which is what removes the compatibility notations.  (I am open to other suggested names; the current names for the flags are because the latest run of `update-compat.py` on a release branch should be with `--release`, and the branch should be just before a run of it with `--master` on the master branch.)

The reason I support this proposal is that it maintains the invariant that `-compat X.X` works with Coq as soon as the version number is bumped to X.X.

I am curious what you think of this proposal (and if you want, I can also update this PR to drop `--master` and `--release` and instead write the explicit version number to the test-suite file).

**Comment by Zimmi48** [comment]
I like your refined proposal very much.

**Comment by JasonGross** [comment]
@Zimmi48
> I like your refined proposal very much.

So is there anything more you want done in this PR, other than fixing the CI?  (I have a preference for letting someone else write the script that bumps version number, etc, in large part because I'm not sure what the arguments should be and what the relevant files to update are.)

**Comment by Zimmi48** [comment]
Let me some time to review this again with a fresh head and put it in the larger context.

**Comment by JasonGross** [comment]
I think the nix packaging job might be bugged.  It's replacing `#!/usr/bin/env bash` with `#!/nix/store/czx8vkrb9jdgjyz8qfksh10vrnqa723l-bash-4.4-p23/bin/bash` in `test-suite/tools/update-compat/run.sh`.  (More discussion on gitter.)  I also want to make `run.sh` read-only; will this break the nix packaging job?

**Comment by ejgallego** [comment]
> `test-suite/tools/update-compat/run.sh`

@JasonGross I think this test is depending on too much hardcoded stuff, I had some headaches with it in dune too. I guess we should find a more robust way of testing what is testing. For example, why is it accessing the docs? IMHO the test suite shouldn't depend on the docs like that.

**Comment by JasonGross** [comment]
@ejgallego It is accessing the docs because:
(a) The list of files under `Compat` in `doc/stdlib/index-list.html.template` must match the currently existing list of actual compat files, and
(b) If the way of doing file organization in `doc/stdlib/index-list.html.template` changes in a way that breaks the compat-update script, then we want to catch this early (when the change to the doc structure is made) rather than late (when it's time to branch the next version)

In some sense, it is testing almost exactly what needs to be tested and no more and no less: it is testing that if you were to bump the version and run the script today, then that process will succeed and produce the correct output.  The "almost" fudge-factor is in assessing what "correct" here means; I take it to mean that if you did not, in fact, bump the version, than the correct output should exactly match the current versions of the files.

**Comment by ejgallego** [comment]
@JasonGross why are you accessing the docs to check that. Check what `.vo` objects are there. Even better, you should know the Compat files, so indeed call `coqtop` to see the files can be required. That's the best option.

Checking the docs is adding a very strong dependency for very little gain, on the verge of technical debt I'd say. For something that is done a couple of years we are already seeing a lot of trouble for such a fresh script. [In the order of hours of dev time]

I ran into this because I am setting up better dependencies for the test-suite. I do not want the test-suite to have to wait for the full `stdlib` doc build, for example; much less i want to deal with this when we replace or improve `coqdoc`.

**Comment by JasonGross** [comment]
@ejgallego I'm a bit confused.  I am not checking the output of the docs.  Morally, this test-suite file depends on every file which is checked in to the Coq source tree, and depends on absolutely no build outputs.  (It would indeed be wrong to check the output of coqdoc.)

Is the issue that dune does dependency granularity on the basis of folders?

**Comment by JasonGross** [comment]
I have updated the compat script to pass through `#!` lines in the bash script, which should hopefully fix the nix packaging job.

**Comment by Zimmi48** [comment]
Good. Note that another possible way would be for bash scripts in the test-suite not have a shebang (and not be executable) and instead be called explicitly with the `bash` command. This would also avoid the shebangs patching.

**Comment by JasonGross** [comment]
@Zimmi48 Status of this?

File: dev/tools/update-compat.py
Comment (Zimmi48) [inline dev/tools/update-compat.py]: Sometimes before branching.

File: dev/tools/update-compat.py
Comment (Zimmi48) [inline dev/tools/update-compat.py]:
```suggestion
  or --master (right after branching and while updating the version number in version.ml)''')
```

**Review by Zimmi48 — COMMENTED** [review]
General LGTM. I mostly have comments on the documentation part.
Sorry for the long delay Jason.

File: dev/tools/update-compat.py
Comment (JasonGross) [reply dev/tools/update-compat.py]: It is important that the version number in `version.ml` is updated *before* running the script.  "while" doesn't convey this to me.

Comment (Zimmi48) [reply dev/tools/update-compat.py]: I see.

File: dev/tools/update-compat.py
Comment (JasonGross) [reply dev/tools/update-compat.py]: It seems a bit less invasive than `git rm`, but there's no strong reason.  Would you prefer just `git rm`?

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Done

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

File: dev/doc/release-process.md
Comment (JasonGross) [reply dev/doc/release-process.md]: Updated

**Comment by JasonGross** [comment]
Sorry @Zimmi48 that it's taken me so long to address your review comments.  I've finally gotten around to rebasing this and updating the doc.  Hopefully this is about ready for merge now.

File: dev/tools/update-compat.py
Comment (Zimmi48) [reply dev/tools/update-compat.py]: I think so yes. Otherwise, you might be left wondering why your tree is not clean.

File: dev/doc/release-process.md
Comment (Zimmi48) [inline dev/doc/release-process.md]: This four-space indentation is interpreted by GitHub rendering engine as meaning a code block. Try using only two spaces.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
I look forward to the next step which will be to run this script in `--release` mode and thus remove the deprecated notations. However, for this we will have to wait for a plan for CompCert to get updated. Given that the primitive projection PR is also hitting compatibility issues with CompCert, let's wait and see.

**Pull request merged**
